### PR TITLE
sensible equals warning: strings and collections 

### PR DESCRIPTION
**Pull request by lrytz**
build-in knowledge about strings and collections
  - strings are only equal to strings
  - collections: equality only within seqs / sets / maps

### CODE REVIEW

**Comment by lrytz** [comment]
Fixes the `List(1) == 1` example from https://github.com/scala/bug/issues/6593#issuecomment-2317245021, but only indirectly by special-casing collections.

**Comment by lrytz** [comment]
@som-snytt would love to hear what you think about my long comment https://github.com/scala/bug/issues/6593#issuecomment-2317245021.

The warn-if-opposite-direction-warns could also be behind Xlint.

File: src/compiler/scala/tools/nsc/typechecker/RefChecks.scala [hunk 1]
```diff
@@ -1042,7 +1042,7 @@ abstract class RefChecks extends Transform {
       // equals.
       def isUsingWarnableEquals = {
         val m = receiver.info.member(nme.equals_)
-        ((m == Object_equals) || (m == Any_equals) || isMethodCaseEquals(m))
+        (m == Object_equals) || (m == Any_equals) || isMethodCaseEquals(m)
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/RefChecks.scala:1045]: the diff shows that you deleted the inner parens. You probably intended to delete the outer parens. (I am trying out some Swiss humour.)

Comment (lrytz) [reply src/compiler/scala/tools/nsc/typechecker/RefChecks.scala:1045]: shoot!

File: src/compiler/scala/tools/nsc/typechecker/RefChecks.scala [hunk 2]
```diff
@@ -1129,7 +1140,21 @@ abstract class RefChecks extends Transform {
...
+        if (!haveSubclassRelationship)
+          nonSensiblyNeq()
+      }
+      else if (isCollection(receiver)) {
+        val rBase = collectionBase(receiver)
+        val aBase = collectionBase(actual)
+        if (rBase != NoSymbol) {
+          if (aBase != NoSymbol && rBase != aBase)
+            nonSensiblyNeq()
+          if (!isCollection(actual) && !haveSubclassRelationship)
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/RefChecks.scala]: `else`

**Review by som-snytt — APPROVED** [review]
Confirmed the new test adds warnings only, for the new string and collection checks.

**Comment by som-snytt** [comment]
`test/files/neg/nonsense_eq_refine.scala` could be moved into one of the canonical tests.

**Comment by som-snytt** [comment]
In case one wonders whether these equality lints are excessive, Intellij announces:
```
New Expression is compared to itself inspection for Java
A new inspection reports cases where an expression is being compared to itself instead of another.
```
image

**Pull request merged**